### ISSUE TITLE
Format document should not add braces by default

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#27821]
Formatting a document by default now has started adding braces around single line blocks:

``` C#
using System;

namespace ConsoleApp75
{
    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length == 0)
                throw new Exception();
        }
    }
}

```

To:

``` C#
using System;

namespace ConsoleApp75
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            if (args.Length == 0)
            {
                throw new Exception();
            }
        }
    }
}

```

However there is no way to enforce whether or not braces are optional for single line statements in editorconfig. This is going to lead to unneeded changes to my tree without a way to enforce it. This option should be turned off by default unless there is a way to enforce it.

**Comment by sharwell** [comment]
I believe the setting is controlled by `csharp_prefer_braces`.

**Comment by davkean** [comment]
Yes that is correct, this works like a charm exactly as I expected. The only question is, should this be the default?

**Comment by davkean** [comment]
I'll change the description.

**Comment by sharwell** [comment]
Marked as **Need Design Review**, but it will *almost certainly* change to **By Design**. The first time this feature is used, it prompts users to configure the items that get formatted, of which braces is one. The setting can be disabled for a team via .editorconfig or a user via Tools &rarr; Options (it only works if both are enabled). This aligns with our design by intent (as opposed to by omission), and thus is unlikely to change unless users are unable to locate the settings which control the feature.

**Comment by sharwell** [comment]
The prompt should show as a gold bar the first time you use Format Document without visiting the C# Formatting page in Tools &rarr; Options.

**Comment by davkean** [comment]
I never got prompted.

**Comment by sharwell** [comment]
This went to a design review today. We ended up deciding to not take action on the original design, and instead focus on ensuring the current design is presented in the manner intended. This specially means making sure features like the gold bar work where intended (https://github.com/dotnet/roslyn/issues/27821#issuecomment-397780098) and the options related to braces get applied in the expected manner following a user change.

In the event we revisit this in the future, we will focus on ensuring the default severity of the "use braces" *diagnostic* matches user expectations. Given the reliability/predictability of the code fix for this diagnostic, we believe it is a reasonable one to include by default in the Code Cleanup feature (candidates for this were identified primarily by their low probability of producing incorrect changes).

**Comment by davkean** [comment]
@sharwell Please point me to the bugs around the gold bar.

**Comment by sharwell** [comment]
> @sharwell Please point me to the bugs around the gold bar.

@JieCarolHu @heejaechang @kuhlenh Can you check internal telemetry to verify users who run Format Document in the latest builds are seeing this bar?

**Comment by davkean** [comment]
Can we please not close this bug until we've got bugs on the gold bar? In my case this was a surprising change of behavior and want to make sure we've captured this in a bug.

**Comment by sharwell** [comment]
@davkean Filed #27974

**Comment by JieCarolHu** [comment]
The gold bar only shows when user click Ctrl+K,D, and the code cleanup is not configured yet

And code cleanup is considered Configured if user goes to the Formatting setting page and click "Ok" to exit.  

One of the scenario that user may never see the gold bar is:
User open the formatting setting page via tools->options without Ctrl+K,D, and click OK. 

I didnot explicitly write any code telemetry for code cleanup, so unless we have some "built-in" tememetry with the gold bar, we probably don't the telemetry...

**Comment by CyrusNajmabadi** [comment]
> And code cleanup is considered Configured if user goes to the Formatting setting page and click "Ok" to exit.

This feels wrong to me.  There is not explicit consent being given when the user just navigates somewhere, then hits 'ok' to close a page.  What would probably make sense would be to change this list:

![image](https://user-images.githubusercontent.com/4564579/41681626-010ac082-748a-11e8-99ce-4a1d1bad0a31.png)

Into something more like:

![image](https://user-images.githubusercontent.com/4564579/41681684-3d2101b2-748a-11e8-90cf-cfceeef79de1.png)

This "perform additional cleanup during formatting" would be unchecked by default.  It would only be checked if the user explicitly clicked it, or if they accepted through the gold-bar.

Thoughts @kuhlenh @sharwell @jinujoseph @JieCarolHu ?

**Comment by CyrusNajmabadi** [comment]
@davkean As well.  I'm curious if you would have preferred the sort of approach i outline above ^

**Comment by kuhlenh** [comment]
Yes, I agree with you Cyrus. Carol, can we still make this change for 15.8?
________________________________

**Comment by heejaechang** [comment]
didn't this opinion show up in design meeting and we said we do this way so that it is more on by default?

I am +1 on off by default for all new features. so I like it being off and require very explicit users opt-in, but I thought our general team policy is we like things to be on by default. and we have gold bar for now since it is experimental? but otherwise, on by default?

**Comment by Neme12** [comment]
I'd personally want this not only off by default for formatting, but as a separate command from formatting. There are many reasons why I'd like to format the document but not cleanup (yet), for example:

1. "make field readonly" I do like to keep my fields readonly, but maybe I just haven't written the code that mutates them yet. I'll do cleanup later (when more of the code is written), but for now, I may want to format whitespace because it is completely broken after some edit.
2. "remove unused variable" same logic - maybe I just haven't written the line that uses this variable yet
3. "remove usings" - again, I could imagine a scenario where I know I will need to use System.Lin, but haven't gotten to that point yet
4. "expression/block body preferences" - same idea, I have written one statement inside a method and I'm about to write another one, but I'd like to do formatting.

All of this makes me worry about whether coupling format and cleanup into the same command is a good idea. (I haven't used this feature yet so I don't know how often this will be a problem in practice, but it's concerns me that if I want to use code cleanup, it needs to always happen together with formatting).

**Comment by CyrusNajmabadi** [comment]
I think 'on by default' and 'off by default' aren't really positions that can be taken without enough context.   it really needs to be considered on a case-by-case basis.

When you have something very large, potential invasive, **and** which takes over an existing key binding people already use (like code-cleanup), then we may have to have a different policy versus a very minor feature that doesn't really impact hte user experience.

I would say that it's more of a sliding scale.  With the more potentially disruptive and contentious features leaning more and more to 'opt in', and the simpler and less invasive stuff being opt-out.  

--

in the case of code-cleaners, it seemed like we wanted to make it opt-in (hence the goldbar), however, we somewhat over-zealously also coded up another form of 'opt-in' that doesn't really match user expectations.  i.e. just opening a dialog, **not** interacting with it, and then hitting "ok" is not an 'opt in' from the mind of most users.  In this scenario (esp. as there was no change), 'ok' is on the same level of 'close' or 'cancel', and shoudl thus not be inferred to be opting in.

--

In the proposal i listed above, we far more clearly can understand the user is opting-in.  This is because they have to explicitly check the "Perform additional code cleanup..." checkbox.  So we can actually understand that this is something they intentionally enabled.

**Comment by heejaechang** [comment]
this is just first step for us to have underlying subsystem and get feedback and incremental design going.

what we probably at the end wants is, not having separate code cleanup feature but giving existing diagnostics, refactoring system more trigger than LB.

right now, only way to apply code fix is LB. 

we want to expand it so that users can define multiple trigger such as format document, edited text span (think like VB line commit), save files, commit, build and etc.

so when people configure analyzer/fixer and its severity, options, they can also configure when fix can be run. right now, it is manual through LB. but user can say analyzer/fixer/fix A should run when I do format document, analyzer/fixer/fix B should run when I commit and etc. and that should include all analyzer/fixer included not just one we picked from builtin analyzer/fixer/fixes.

and also provide a new batch fix feature which users can select all analyzer/fixer/fixes they want to apply to a scope (projects, solution and etc) with custom options and bulk apply those changes.

we might also provide code fix from error list.

anyway, we welcome any feedbacks!

**Comment by CyrusNajmabadi** [comment]
> anyway, we welcome any feedbacks!

Definitely!

I think we're def at a decent starting position, and i'm glad that we've been able to ship *something* for people to be able to use and respond to.  I also think there's a wealth of improvements we can make here (most of which you touched on yourself in your post).  So i def hope we'll continue to improve this area greatly in the coming months!

**Comment by Neme12** [comment]
> we want to expand it so that users can define multiple trigger such as format document, edited text span (think like VB line commit), save files, commit, build and etc.

That sounds interesting and powerful but will there ever be a way to run code cleanup *without* having to bind it to any of these other commands? Thanks

> but user can say analyzer/fixer/fix A should run when I do format document, analyzer/fixer/fix B should run when I commit and etc. 

Can you help me understand what's the use case for this? What would be an example of A and B such that someone would like to run A on format but B on save. Is that much flexibility actually useful as opposed to being overwhelming to the user? Do we expect most users to tinker with these individually? (Maybe there is a compelling use case for that, that's why I'm asking for an example so that I can see this better. Thanks)

What I had in mind is exactly what @CyrusNajmabadi said - Code cleanup would be a separate command and there would be a page in settings to configure what that entails - which features get run during code cleanup. Then there would be on/off options to run code cleanup 1) during formatting 2) during save 3) during build? etc. But this would of course remove the ability to run a different set of cleanup features with different triggers.

**Comment by heejaechang** [comment]
we would like to think it as a code transformation which are provided by diagnostic/fixer or refactoring. and has options around that. 

we do not want to have 2 features that basically do same thing. then we get into confusion where options for 1 feature (diagnostic analyzer) applying to another feature (code cleanup). 

for example, if add brace diagnostic analyzer is disabled by ruleset editor, should it run on code clean up feature?

that is why we want to consolidate all those in 1 place where users set all options including when code tranformation runs. and not think these as separate features.

…

and sure we can add new command if user want rather than merging it to format document.

…

the reason we have multiple triggers is that, if this is generic mechanism to run multiple code transformation including third party analyzer, it might take long time to do everything.

so, you might want to do only faster and quick stuff while format document but when you commit or save, you are willing to spend some more time to clean more things up.

at least, that was one of feedback from diagnostic/fixer side. such as I have bunch of errors/suggestion from analyzers, I want all those to be automaticallly fixed up when I commit/save.

that is why we have save/commit as one of trigger.

**Comment by heejaechang** [comment]
tagging @kuhlenh @jinujoseph since all these are design question.

**Comment by sharwell** [comment]
The check box requested in https://github.com/dotnet/roslyn/issues/27821#issuecomment-398878523 already exists in the UI. It just has visibility set to collapsed (not visible).

**Comment by Neme12** [comment]
> it might take long time to do everything

This is also another reason why I think I'd prefer a separate command for cleanup and not bind this to format, save or build.

**Comment by davkean** [comment]
> it might take long time to do everything

Not might, it does take a long time: https://github.com/dotnet/roslyn/issues/28012.

**Comment by CyrusNajmabadi** [comment]
>>     it might take long time to do everything
> This is also another reason why I think I'd prefer a separate command for cleanup and not bind this to format, save or build.

Note: i've collected data on this, and it's not pretty: https://github.com/dotnet/roslyn/pull/28029#issuecomment-398925871

TLDR: currently (without a PR i'm working on), it can take >2.5 minutes to just format-document on a file in Roslyn.  Even with the PR, we're still at around 100 seconds to format.  This is *super brutal* and is likely never doing to be acceptable in perf-sensitive scenarios like save.  

Now, we can def try to make all the cleaners faster (and i'm trying to figure out how i can capture that information myself to help speed these up).  But it's likely that we're always going to have some amount of fixed cost here that makes these times very unpalatable for some users.

**Comment by kuhlenh** [comment]
We have a "north star" vision that includes something like an additional "Apply EditorConfig" command and a more flexible code cleanup. For 15.8, we are putting together this MVP to get feedback and verify developer scenarios.

I do agree with Cyrus that we should probably have that top-level check box that is "Perform additional code cleanup during formatting". To my knowledge, (@JieCarolHu can confirm), all "additional" cleanup options are off-by-default *unless* you hit "Configure" from the gold bar. Then we will turn on that set of default options.

Thank you for all this feedback--would love to hear about your experience with 15.8 p3's prototype so we can gather all the right scenarios to perfect our north star design. Feel free to email me your feedback from this experience at kaseyu@microsoft.com

**Comment by davkean** [comment]
> all "additional" cleanup options are off-by-default unless you hit "Configure" from the gold bar. T

That is not my experience, hence the above bug.

**Comment by heejaechang** [comment]
@davkean we enabled the only quick one by default and left expansive one there but off by default. if you enabled them all, then sure it will take some time to run. especially if there are a lot of changes to make.

**Comment by davkean** [comment]
Which comment are you referring to? The speed? Or the defaults?

**Comment by heejaechang** [comment]
options in format document

![image](https://user-images.githubusercontent.com/1333179/41690993-0e7774b8-74ad-11e8-8356-86b515253a0f.png)

only cheap one is on by default. and all expansive ones are off by default. if they got all on by default, that is some other bug.

**Comment by heejaechang** [comment]
by the way, these are roaming options. meaning, if you set those in one VS, all your other VS including different hives will all get that.

**Comment by davkean** [comment]
@heejaechang Please read the title of the bug, this bug is about that add/remove braces is on by default or results in a behavior change.

**Comment by heejaechang** [comment]
@davkean sorry. I am mixing this https://github.com/dotnet/roslyn/issues/28012 with this.

I think for this particular one, we decided to turn off "add brace" by default, @kuhlenh

**Comment by heejaechang** [comment]
alright, I might misunderstood. leaving it to @kuhlenh whether "add brace" is on by default or not.

I think there are 3 different things here.

1. whether code clean up gets enabled without user explicit opt-in through gold bar + format option page. (it should off until explicitly opted in, otherwise, it is a bug)
2. whether "add brace" feature should be on by default in code clean or not. (@kuhlenh )
3. perf of code clean up. (right now, we only turned on ones that we know fast by default but still put expansive one there in case users want those)

**Comment by kuhlenh** [comment]
This bug is closed as "by design" per what @sharwell indicated we decided in a design meeting. 

@davkean I'm going to address the default config issue you have in #27974. As @heejaechang mentioned, the config is off unless you opt-in. I will discuss with you on this other issue.

**Comment by heejaechang** [comment]
actually, there are more things like

whether we want an explicit command for code clean or not. or we want separate code clean up features altogether or not. but all those are something part of longer design decisions we need to make.

**Comment by davkean** [comment]
I think the disconnect is the act "opting-in" which we can discuss under: https://github.com/dotnet/roslyn/issues/27974.
